### PR TITLE
Fix bad discount for virtual product (apply fix by @nziria)

### LINKED ISSUE #10264: Major bug with Virtual Products
Hello folks,

I think I found a major bug in the handling of Virtual Products in Prestashop 1.7.

First off some background:

A while ago, I asked for help with an issue I have about the total being incorrectly calculated when using discount vouchers:

https://www.prestashop.com/forums/topic/866094-total-doesnt-consider-voucher/

As you can see, no one bothered to give me an answer as of now. Because this issue keep costing me money I kept investigating. I set up a test server using the latest Prestashop build 1.7.4.2 and installed my theme to it, but other than that didn't bother to add any modules or even products. I found that with the default products the discount voucher was displayed correctly:

https://i.imgur.com/xHBmxKR.jpg

After a lot of testing I found the culprit! When I changed the product type to "Virtual Product" the discount voucher stopped being calculated correctly:

https://i.imgur.com/n7wdC7d.jpg

https://i.imgur.com/HuRWDFw.jpg

I considered that this might be an issue related to my custom theme, but even using "Classic" that comes with 1.7.4.2 the problem still remained.

I really hope you could look into this matter and fix it for the next update and it would be great if someone could provide me some clues how I can remedy the problem in the meantime. As I mentioned, I am actually missing out on sales right now, because people can't use their discount codes.

Cheers

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Apply fix suggested by @nziria on https://github.com/PrestaShop/PrestaShop/issues/10264#issuecomment-458211709 which fixes discount application to virtual products
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/10264#issuecomment-458211709
| How to test?  | See reported bug

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12384)

### CODE REVIEW

**Review by PierreRambaud — APPROVED** [review]

File: classes/Cart.php [hunk 1]
```diff
@@ -1914,7 +1914,7 @@ public function getOrderTotal(
             return 0;
         }
         if ($virtual && $type == Cart::BOTH) {
-            $type = Cart::BOTH_WITHOUT_SHIPPING;
+            $type = Cart::BOTH;
```
Comment (Quetzacoalt91) [inline classes/Cart.php:1917]: The whole `if` statement is actually unnecessary.

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
The boss said: NOPE

**Comment by Quetzacoalt91** [comment]
In my opinion this change is not enough to be approved. It seems to solve an issue to someone without breaking the existing Travis builds, but this case must be added to the tests.

@tomlev made a huge work to make this method tested, so why not trying to add a case for virtual product in `CartGetOrderTotalTest`?

https://github.com/PrestaShop/PrestaShop/blob/c33d60e02e1b3270bd192aadfd85ae7a31dd562f/tests-legacy/Integration/classes/CartGetOrderTotalTest.php#L50

**Comment by tomlev** [comment]
the right folder to add tests to this method is `/tests-legacy/Unit/Core/Cart/Calculation/`
you can use any directory within it as an example ;)
let me know if you need some help

File: classes/Cart.php (hunk 1 above)
Comment (pandisoft) [reply classes/Cart.php:1917]: The fix works but if the discount is 100% you can not place the order, the button "Order with an obligation to pay" still appear and disabled 
https://image.prntscr.com/image/viS6zjKXQJCoX8h1uAV85g.png

Comment (matks) [reply classes/Cart.php:1917]: Hi @pandisoft thanks for the feedback.

I'm going to add a few unit tests following @Quetzacoalt91 recommandations in order to cover all of these usecases so we know we're in safe waters 😄 I'll add this usecase then

**Comment by matks** [comment]
Will be fixed by tomlev in https://github.com/PrestaShop/PrestaShop/pull/12818

**Comment by matks** [comment]
Will be fixed in https://github.com/PrestaShop/PrestaShop/pull/12965 finally 😄 

@pandisoft would you please have a look at https://github.com/PrestaShop/PrestaShop/pull/12965 to check whether this solves your usecase, or the bug you reported is still there ?

**Pull request closed without merging**